File: misc/valid-sudoku.c
bool isValidSudoku(char** board, int row, int col) {
    int i,j;
    int key;
    int mem[10];

    for(i = 0; i < row; ++i){
        memset(mem,0,sizeof(int) * 10);
        for(j = 0; j < col; ++j){
            if(board[i][j] == '.')
                continue;
            key = board[i][j] - '0';
            if(mem[key] > 0)
                return false;
            mem[key]++;
        }
    }

    for(j = 0; j < col; ++j){
        memset(mem,0,sizeof(int)*10);
        for(i = 0; i < row; ++i){
            if(board[i][j] == '.')
                continue;
            key = board[i][j] - '0';
            if(mem[key] > 0)
                return false;
            mem[key]++;
        }
    }

    for(i = 0; i < 9; ++i){
        memset(mem,0,sizeof(int) * 10);
        for(j = 0; j < 9; ++j){
            int x = (i / 3) * 3 + (j / 3);
            int y = (i % 3) * 3 + (j % 3);

            if(board[x][y] == '.')
                continue;
            key = board[x][y] - '0';
            if(mem[key] > 0)
                return false;
            mem[key]++;
        }
    }

    return true;
}

File: python/hash-table/valid-sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        for i in range(len(board)):
            d = {}
            for j in range(len(board[i])):
                if board[i][j] == '.':
                    continue
                if board[i][j] in d:
                    return False
                d[board[i][j]] = 1

        for j in range(len(board[0])):
            d = {}
            for i in range(len(board)):
                if board[i][j] == '.':
                    continue
                if board[i][j] in d:
                    return False
                d[board[i][j]] = 1

        for i in range(9):
            d = {}
            for j in range(9):
                row = (i // 3) * 3 + j // 3
                col = (i % 3) * 3 + j % 3
                if board[row][col] == '.':
                    continue
                if board[row][col] in d:
                    return False
                d[board[row][col]] = 1

        return True
